terraform/gcp.tf

```diff
@@ -16,19 +16,32 @@ resource "random_id" "random" {
   byte_length = "8"
 }
 
-# Create the project
+# Create the project if one isn't specified
 resource "google_project" "vault" {
+  count           = "${var.project != "" ? 0 : 1}"
   name            = "${random_id.random.hex}"
   project_id      = "${random_id.random.hex}"
   org_id          = "${var.org_id}"
   billing_account = "${var.billing_account}"
 }
 
+# Or use an existing project, if defined
+data "google_project" "vault" {
+  count           = "${var.project != "" ? 1 : 0}"
+  project_id      = "${var.project}"
+}
+
+# Obtain the project_id from either the newly created project resource or existing data project resource
+# One will be populated and the other will be null
+locals {
+  vault_project_id = "${join(",",data.google_project.vault.*.project_id)}${join(",",google_project.vault.*.project_id) }"
```

Comment:
I think you can use a ternary operator here too instead of the string join, which might make it clearer

Reply:
In my first pass, I tried used a ternary operator here.  It fails out of the gate because it doesn't handle one of the resource references not being present (see: https://stackoverflow.com/a/41640684).  In this case, one of the two will always not be there. The join approach, while not as "clean", handles this somewhat gracefully.

Reply:
A colleague pointed me to the element/concat approach which I just dropped in.